book-content/chapters/14-configuring-typescript.md

```diff
@@ -0,0 +1,1024 @@
+
+## ESM and CommonJS
+
+There are two ways to modularize your code in TypeScript: ECMAScript Modules (ESM) and CommonJS (CJS). These two module systems operate slightly differently, and they don't always work together cleanly.
+
+ES Modules use `import` and `export` statements:
+
+```typescript
+import { createAlbum } from "./album";
+
+export { createAlbum };
+```
+
+CommonJS uses `require` and `module.exports`:
+
+```typescript
+const { createAlbum } = require("./album");
+
+module.exports = { createAlbum };
+```
+
+Understanding the interoperability issues between ESM and CJS is a little beyond the scope of this book. Instead, we'll look at how to set up TypeScript to make working with both module systems as easy as possible.
+
+### How Does TypeScript Know What Module System To Emit?
+
+Imagine we have our `album.ts` file that exports a `createAlbum` function:
+
+```typescript
+// album.ts
+
+export function createAlbum(
+  title: string,
+  artist: string,
+  year: number,
+): Album {
+  return { title, artist, year };
+}
+```
+
+When this file is turned into JavaScript, should it emit `CJS` or `ESM` syntax?
+
+```javascript
+// ESM
+
+export function createAlbum(title, artist, year) {
+  return { title, artist, year };
+}
+```
+
+```javascript
+// CJS
+
+function createAlbum(title, artist, year) {
+  return { title, artist, year };
+}
+
+module.exports = {
+  createAlbum,
+};
+```
+
+The way this is decided is via `module`. You can hardcode this by choosing some older options. `module: CommonJS` will always emit CommonJS syntax, and `module: ESNext` will always emit ESM syntax.
+
+But if you're using TypeScript to transpile your code, I recommend using `module: NodeNext`. This has several complex rules built-in for understanding whether to emit CJS or ESM:
+
+The first way we can influence how TypeScript emits your modules with `module: NodeNext` is by using `.cts` and `.mts` extensions.
+
+If we change `album.ts` to `album.cts`, TypeScript will emit CommonJS syntax, and the emitted file extension will be `.cjs`.
+
+If we change `album.ts` to `album.mts`, TypeScript will emit ESM syntax, and the emitted file extension will be `.mjs`.
+
+If we keep `album.ts` the same, TypeScript will look up the directories for the closest `package.json` file. If the `type` field is set to `module`, TypeScript will emit ESM syntax. If it's set to `commonjs` (or unset, matching Node's behavior), TypeScript will emit CJS syntax.
+
+| File Extension | Emitted File Extension | Emitted Module System               |
+| -------------- | ---------------------- | ----------------------------------- |
+| `album.mts`    | `album.mjs`            | ESM                                 |
+| `album.cts`    | `album.cjs`            | CJS                                 |
+| `album.ts`     | `album.js`             | Depends on `type` in `package.json` |
```

Comment:
I think it's worth emphasizing this again here:
```suggestion
| `album.ts`     | `album.js`             | Depends on `type` in the closest `package.json` |
```

Nowadays it's not entirely uncommon to see such package.json files sprinkled in codebases here and there:
```ts
{ "type": "commonjs" }
```

They look... weird but they are totally legal. You can use such package.json files to "switch" the default type of .js/.ts for a certain directory.